timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,784 @@
+            restraint_cross_angle_params.append(1000.0)
+        else:
+            assert 0, "Illegal Geometry"
+    elif anchor_core_geometry == LocalGeometry.G2_LINEAR:
+        assert 0, "Illegal Geometry"
+    elif anchor_core_geometry == LocalGeometry.G3_PLANAR:
+        assert 0, "Illegal Geometry"
+    elif anchor_core_geometry == LocalGeometry.G3_PYRAMIDAL:
+        j = anchor
+        if anchor_full_geometry == LocalGeometry.G4_TETRAHEDRAL:
+            #            c
+            #           /
+            # type a - j . k
+            #           \
+            #            b
+            #
+            # we have some choices here:
+            # 1) if there is no ring-opening and closing, then we use a centroid angle restraint defined by [a,b,c],j,k
+            # 2) if there is ring-opening and closing, then we need to enumerate possible cross-product based restraints.
+            # we currently support only 1) right now. But it would not be difficult to implement 2).
+
+            # it turns out that 1) is quite difficult, since ammonium groups do not have reasonable centroids due to
+            # rapid conversion. So we should always do 2) as it's more robust and enables more kinds of transformations.
+
+            a, b, c = nbs_1
+
+            # enumerate all possible anchors, break when we find a stable one
+            x, y = None, None
+            for xx, yy in [(a, b), (a, c), (b, c)]:
+
+                b_okay = check_bond_angle_stability(
+                    xx, j, yy, mol_b_bond_idxs, mol_b_bond_params, mol_b_angle_idxs, mol_b_angle_params
+                )
+                a_okay = check_bond_angle_stability(
+                    core_b_to_a[xx],
+                    core_b_to_a[j],
+                    core_b_to_a[yy],
+                    mol_a_bond_idxs,
+                    mol_a_bond_params,
+                    mol_a_angle_idxs,
+                    mol_a_angle_params,
+                )
+
+                if b_okay and a_okay:
+                    x = xx
+                    y = yy
+                    break
+
+            atoms = find_attached_dummy_atoms(dg, mol_b_bond_idxs, anchor)
+            assert len(atoms) == 1
+            k = atoms[0]
+
+            restraint_cross_angle_idxs.append(((j, x), (j, y), (j, k)))
+            restraint_cross_angle_params.append(1000.0)
+
+        else:
+            assert 0, "Illegal Geometry"
+
+    else:
+        assert 0, "Illegal Geometry"
+
+    all_idxs = (
+        restraint_bond_idxs,
+        restraint_angle_idxs,
+        restraint_proper_idxs,
+        restraint_improper_idxs,
+        restraint_cross_angle_idxs,
+        restraint_centroid_angle_idxs,
+    )
+    all_params = (
+        restraint_bond_params,
+        restraint_angle_params,
+        restraint_proper_params,
+        restraint_improper_params,
+        restraint_cross_angle_params,
+        restraint_centroid_angle_params,
+    )
+    return all_idxs, all_params
```

Comment:
probably! would a named tuple suffice?

Reply:
willfix - but not sure on the best way to do it yet, may just stuff them onto lib.Potentials objects

Reply:
Does something like the solution in https://github.com/proteneer/timemachine/pull/746/commits/6d68a980d21b1e191038cf7ea8142dc2455fe4fc work?

Reply:
Nice -- this looks like an improvement!

Mainly want to group together the idxs and params for each restraint type. Later it might be worthwhile to extract more of the validation logic into each restraint type's constructor / updater methods, or into the constructor of `VacuumSystem`, as appropriate.